### PR TITLE
[V2] tpm2_import: add -A option to specify the object attributes

### PR DESCRIPTION
**Pull request by jiazhang0**
Signed-off-by: Jia Zhang <qianyue.zj@alibaba-inc.com>

### CODE REVIEW

File: tools/tpm2_import.c [hunk 1]
```diff
@@ -51,6 +51,7 @@
 #include "tpm_kdfa.h"
 #include "tpm2_options.h"
 #include "tpm2_util.h"
+#include "tpm2_tool.h"
```
Comment (martinezjavier) [inline tools/tpm2_import.c]: Can you please keep the headers sorted alphabetically? So this should be after tpm2_options.h.

File: tools/tpm2_import.c [hunk 2]
```diff
@@ -83,6 +84,10 @@ struct tpm_import_ctx {
     TPM2B_DATA enc_sensitive_key;
     TPM2B_MAX_BUFFER encrypted_inner_integrity;
     TPM2B_MAX_BUFFER encrypted_duplicate_sensitive;
+    UINT32 objectAttributes;
```
Comment (martinezjavier) [inline tools/tpm2_import.c]: Why do you need this variable? It isn't used AFAICT.

File: tools/tpm2_import.c [hunk 3]
```diff
@@ -83,6 +84,10 @@ struct tpm_import_ctx {
...
+    struct {
+        UINT16 A : 1;
+    } flags;
```
Comment (martinezjavier) [inline tools/tpm2_import.c]: I don't think this flag is needed either, you can just check if ctx.objectAttributes != 0

File: tools/tpm2_import.c [hunk 4]
```diff
@@ -288,6 +293,13 @@ static bool create_import_key_public_data_and_name(void) {
 
     IMPORT_KEY_SYM_PUBLIC_AREA(ctx.import_key_public)
 
+    if (ctx.flags.A) {
+        ctx.import_key_public.t.publicArea.objectAttributes.val = ctx.objectAttributes;
+    }
+
+    tpm2_tool_output("ObjectAttribute: 0x%08X\n",
+                     ctx.import_key_public.t.publicArea.objectAttributes.val);
+
```
Comment (martinezjavier) [inline tools/tpm2_import.c]: I think this should only be printed if ctx.objectAttributes was assigned a value.

File: tools/tpm2_import.c (hunk 2 above)
Comment (jiazhang0) [reply tools/tpm2_import.c]: It is used to retrieve the value of -A option from commandline:

```
 +    case 'A':
 +        if(!tpm2_util_string_to_uint32(value, &ctx.objectAttributes)) {
 +            LOG_ERR("Invalid object attribute, got\"%s\"", value);
 +            return false;
 +        }
 +        ctx.flags.A = 1;
 +        break;
```

File: tools/tpm2_import.c (hunk 3 above)
Comment (jiazhang0) [reply tools/tpm2_import.c]: This design refers to tpm2_create.c. Yes, if we have -A option only, it is not necessary to use  flags.

File: tools/tpm2_import.c (hunk 4 above)
Comment (jiazhang0) [reply tools/tpm2_import.c]: This also refers to tpm2_create and tpm2_createprimary.

**Comment by jiazhang0** [comment]
V2 Changelog:
- minor changes according to martinezjavier's comments.
- here is a quick diff against V1:
```
diff --git a/tools/tpm2_import.c b/tools/tpm2_import.c
index ae83aa7..dfdd4ae 100644
--- a/tools/tpm2_import.c
+++ b/tools/tpm2_import.c
@@ -50,8 +50,8 @@
 #include "files.h"
 #include "tpm_kdfa.h"
 #include "tpm2_options.h"
-#include "tpm2_util.h"
 #include "tpm2_tool.h"
+#include "tpm2_util.h"
 
 #define SYM_KEY_SIZE 16
 #define max_buffer_size  1024
@@ -85,9 +85,6 @@ struct tpm_import_ctx {
     TPM2B_MAX_BUFFER encrypted_inner_integrity;
     TPM2B_MAX_BUFFER encrypted_duplicate_sensitive;
     UINT32 objectAttributes;
-    struct {
-        UINT16 A : 1;
-    } flags;
 };
 
 static tpm_import_ctx ctx = { 
@@ -293,7 +290,7 @@ static bool create_import_key_public_data_and_name(void) {
 
     IMPORT_KEY_SYM_PUBLIC_AREA(ctx.import_key_public)
 
-    if (ctx.flags.A) {
+    if (ctx.objectAttributes) {
         ctx.import_key_public.t.publicArea.objectAttributes.val = ctx.objectAttributes;
     }
 
@@ -570,7 +567,6 @@ static bool on_option(char key, char *value) {
             LOG_ERR("Invalid object attribute, got\"%s\"", value);
             return false;
         }
-        ctx.flags.A = 1;
         break;
     default:
         LOG_ERR("Invalid option");
```

File: tools/tpm2_import.c (hunk 2 above)
Comment (martinezjavier) [reply tools/tpm2_import.c]: ah, sorry. I misread and thought you were adding both a struct tpm_import_ctx member and a global variable.

**Pull request merged**